### PR TITLE
[BEAM-625] Making Dataflow Python Materialized PCollection representation more efficient (3 of several).

### PR DESCRIPTION
**Pull request by katsiapis**
- Refactoring up common functionality from TextFileSink to FileSink.

- Removing no-longer needed overrides from TextFileSink
  (since their implementation is identical to FileSink).

- Introducing NativeFileSink and NativeFileSinkWriter that capture the common
  patterns of file-related native sinks.

- NativeTextFileSink now inherits from NativeFileSink.

- TextFileWriter now inherits from NativeFileWriter.

- Introducing NativeFileSource and NativeFileSourceWriter that capture the common
  patterns of file-related native sources.

- TextFileSource now inherits from NativeFileSource.

- TextFileReader now inherits from NativeFileSourceReader.

- Introducing CompressionTypes.AUTO, in analogy to Dataflow Java's TextIO.CompressionType.
  This is now the new default for NativeFileSource, NativeFileSink, FileSink (and all
  their subclasses).

- Moving compression_type and mime_type handling to ChannelFactory.open().

- Changing compression level from 9 to Z_DEFAULT_COMPRESSION as that better matches
  disk/cpu tradeoffs.

- Removing CompressionTypes.DEFLATE since it's not a standard on-disk format (and not
  recognized by the service).

- Augmenting the functionality of _CompressedFile and GcsBufferedWriter (supporting more
  standard file operations on them, needed by various Sources/Sinks).

### CODE REVIEW

**Comment by katsiapis** [comment]
Hi @chamikaramj  can you please take a look?

File: sdks/python/apache_beam/io/fileio.py [hunk 1]
```diff
@@ -34,17 +33,301 @@
...
+    self.seek_to_true_start_offset()
+
+    return self
+
+  def __exit__(self, exception_type, exception_value, traceback):
+    self.file.close()
+
+  def __iter__(self):
+    if self.current_offset > 0 and ChannelFactory.is_compressed(self.file):
+      # TODO: Turns this warning into an exception soon.
```
Comment (chamikaramj) [inline sdks/python/apache_beam/io/fileio.py]: Let's add a comment here with something like.

When compression is enabled both initial and dynamic splitting should be prevented. Here we prevent initial splitting by ignoring all splits other than the split that starts at byte 0.

File: sdks/python/apache_beam/io/fileio.py [hunk 2]
```diff
@@ -34,17 +33,301 @@
...
+    """
+    raise NotImplementedError
+
+  def get_progress(self):
+    return iobase.ReaderProgress(position=iobase.ReaderPosition(
+        byte_offset=self.range_tracker.last_record_start))
+
+  def request_dynamic_split(self, dynamic_split_request):
+    if ChannelFactory.is_compressed(self.file):
+      # TODO: Turns this warning into an exception soon.
```
Comment (chamikaramj) [inline sdks/python/apache_beam/io/fileio.py]: Let's add a comment here with something like.

When compression is enabled both initial and dynamic splitting should be prevented. Here we prevent dynamic splitting by ignoring all dynamic split requests at the reader.

File: sdks/python/apache_beam/io/fileio_test.py [hunk 3]
```diff
@@ -78,6 +91,88 @@ def test_read_entire_file(self):
...
 
+  def test_read_entire_file_empty(self):
+    source = fileio.TextFileSource(file_path=self.create_temp_file(''))
+    read_lines = []
+    with source.reader() as reader:
+      for line in reader:
+        read_lines.append(line)
+    self.assertEqual(read_lines, [])
+
+  def test_read_entire_file_gzip(self):
```
Comment (chamikaramj) [inline sdks/python/apache_beam/io/fileio_test.py]: Can you add a test that tries to read a TextFileSource starting at a offset >0 and ensures that nothing gets read ?

File: sdks/python/apache_beam/io/fileio_test.py [hunk 4]
```diff
@@ -78,6 +91,88 @@ def test_read_entire_file(self):
...
+    lines = ['First', 'Second', 'Third']
```
Comment (chamikaramj) [inline sdks/python/apache_beam/io/fileio_test.py]: Probably let's also add a read test for each compression type that reads a larger input (1k records ?)

File: sdks/python/apache_beam/io/fileio_test.py [hunk 5]
```diff
@@ -78,6 +91,88 @@ def test_read_entire_file(self):
...
+    self.assertEqual(read_lines, [])
+
+  def test_read_entire_file_zlib(self):
+    lines = ['First', 'Second', 'Third']
+    compressor = zlib.compressobj(-1, zlib.DEFLATED, zlib.MAX_WBITS)
+    data = compressor.compress('\n'.join(lines)) + compressor.flush()
+    source = fileio.TextFileSource(
+        file_path=self.create_temp_file(data),
+        compression_type=fileio.CompressionTypes.ZLIB)
+    read_lines = []
```
Comment (chamikaramj) [inline sdks/python/apache_beam/io/fileio_test.py]: Let's add a compressed source read test that reads a source for a range [0, b) when the size of file is 'c' where c > b and makes sure that we read the full range [0, c).

**Comment by chamikaramj** [comment]
LGTM

File: sdks/python/apache_beam/io/fileio.py (hunk 1 above)
Comment (katsiapis) [inline sdks/python/apache_beam/io/fileio.py]: Done.

File: sdks/python/apache_beam/io/fileio.py (hunk 2 above)
Comment (katsiapis) [inline sdks/python/apache_beam/io/fileio.py]: Done.

File: sdks/python/apache_beam/io/fileio_test.py (hunk 3 above)
Comment (katsiapis) [inline sdks/python/apache_beam/io/fileio_test.py]: Done.

File: sdks/python/apache_beam/io/fileio_test.py (hunk 4 above)
Comment (katsiapis) [inline sdks/python/apache_beam/io/fileio_test.py]: Done (10K records).

File: sdks/python/apache_beam/io/fileio_test.py (hunk 5 above)
Comment (katsiapis) [inline sdks/python/apache_beam/io/fileio_test.py]: Done.

**Comment by katsiapis** [comment]
Thanks Cham, all comments addressed, PTAL!

**Comment by chamikaramj** [comment]
Thanks.

LGTM.

**Comment by katsiapis** [comment]
Thanks @chamikaramj.
What's the process now? Do we direct to @dhalperi for final incoproration?

**Comment by chamikaramj** [comment]
Yeah. This has to be merged by a committer.

**Comment by katsiapis** [comment]
R: @dhalperi 
Dan, could you take a look?

**Comment by dhalperi** [comment]
LGTM, squashed and merged ; close PR at your leisure.

**Comment by katsiapis** [comment]
Thanks Dan, much appreciated!

Closing PR...

**Pull request closed without merging**